k2/csrc/thread_pool.h

```diff
@@ -0,0 +1,108 @@
+/**
+ * @brief Thread pool.
+ *
+ *
+ * @copyright
+ * Copyright (c)  2020  Xiaomi Corporation (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+#ifndef K2_CSRC_THREAD_POOL_H_
+#define K2_CSRC_THREAD_POOL_H_
+
+#include <condition_variable>  // NOLINT
+#include <functional>
+#include <mutex>  // NOLINT
+#include <queue>
+#include <thread>  // NOLINT
+#include <vector>
+
+#include "k2/csrc/log.h"
+
+namespace k2 {
+
+class ThreadPool {
+ public:
+  using TaskFunc = std::function<void()>;
+
+  /* Create a pool with given number of threads.
+   *
+   * @param [in] num_threads  The number of threads in the pool. If it is <= 0,
+   *                          the number of threads in the pool is set to
+   *                          `std::thread::hardware_concurrency()`.
+   */
+  explicit ThreadPool(int32_t num_threads);
+
+  /* It ask all threads in the pool to exit after processing what they
+   * are currently executing. If there are still tasks in the queue, they
+   * will not be executed.
+   */
+  ~ThreadPool();
+
+  /* Insert a task into the queue. It is executed immediately
+   * if there are currently idle threads. Otherwise, it is saved
+   * in the queue and waits for threads to be free to pop it from
+   * the queue and to execute it.
+   *
+   * @param [in] task   The task to be run. It should be convertible to
+   *                    `std::function<void()>`.
```

Comment:
Not sure if we really need to support other function type, but it's doable with parameter pack and `future`, e.g.

```
  template<typename F, typename...Args>
  auto SubmitTask(F&& f, Args&&... args) -> std::future<decltype(f(args...))> {
//  bind function with parameters
    std::function<decltype(f(args...))()> fun= std::bind(std::forward<F>(f), std::forward<Args>(args)...);
// wrap to a shared_ptr so we can copy in below lambda
    auto task_ptr = std::make_shared<std::packaged_task<decltype(f(args...))()>>(fun);
   // convert to `std::function<void()>`
    std::function<void()> wrapper_fun = [task_ptr]() {
      (*task_ptr)(); 
    };
   // push to taks_ and notify
  return task_prt->get_future();
```
Actually this is a pattern to work with C++ threads.

Reply:
Do we need the return value of a task in k2?